README.md
# Dropbox Backup

`Dropbox Backup` is a collection of **BASH** scripts for backing up your server with [Dropbox Uploader](https://github.com/andreafabrizi/Dropbox-Uploader/).

## Setup

1. First clone this repo:
```bash
$ cd ~
$ git clone https://github.com/ClickBeyond/DropboxBackup.git
```

2. Then give the scripts execution permissions if they don't already have them:
```bash
$ cd ~/DropBoxBackup
$ chmod +x dropbox_uploader.sh
$ chmod +x db_backup.sh
$ chmod +x log_backup.sh
$ chmod +x web_cron.sh
```

3. Finally, run the `dropbox_uploader.sh` script and follow the on-screen instructions to connect it to your Dropbox account:
```bash
$ ./dropbox_uploader.sh
```

## MySQL DB Backup Usage

Make sure the MySQL Client is installed for your version of MySQL Server:
```bash
$ sudo apt-get update
$ sudo apt-get install mysql-client-5.6	# replace 5.6 with your version of MySQL
```

Use the `mysql_config_editor` (new in v5.6.6) to store the database login credentials in an encrypted format:
```bash
$ mysql_config_editor set --login-path=loginPathName --host=hostname --user=username --password
```
Replace `loginPathName` with a custom name and replace `hostname` and `username` with your login credentials. You will be prompted to enter your password.

Some useful `mysql_config_editor` commands are:
```bash
$ mysql_config_editor print --all	# to view the stored login credentials (passwords are displayed as *****)
$ mysql_config_editor remove --login-path=loginPathName	# to remove a stored login
```

The `loginPathName` name is used when backing up MySQL databases. The syntax is:

```bash
$ ~/DropboxBackup/db_backup.sh -l loginPathName -d dbName
```

This script will create a MySQL dump file, which is then compressed into a `.tar.gz` file and then uploaded to your dropbox folder.

## Backup Log File

Every time the `db_backup.sh` script is executed, entries are added to `~/tmp/backup.log`. This log file can be backed up to dropbox by executing the `log_backup.sh` script as follows:

```bash
$ ~/DropboxBackup/log_backup.sh
```

Alternatively, the log file can be backed up manually by doing the following:
```bash
$ cd ~/tmp
$ BKP_LOG_FILE="log-backup-$(date +"%Y-%m-%d_%H-%M-%S").tar.gz"
$ tar -zcf "$BKP_LOG_FILE" "backup.log"
$ ~/DropboxBackup/dropbox_uploader.sh -f ~/.dropbox_uploader upload $BKP_LOG_FILE "/Log_Backups/$BKP_LOG_FILE"
$ rm -f $BKP_LOG_FILE
```

The above commands perform the following:

1. Change to the `tmp` directory.
2. Create a new name for the backup.
3. Compress the existing log file into a `.tar.gz`.
4. Upload the compressed file to dropbox.
5. Delete the archive.

## Scheduled Jobs (cron)

Cron is used on Linux to schedule commands or scripts that need to be executed periodically.

To edit your `crontab` file use:
```bash
$ crontab -e
```

Providing your system is setup to send emails (see [Sending Email using Mandrill](#6)), then add these lines to your `crontab` file to email errors to the `MAILTO` address:
```
MAILTO="<your@email.com>"
@daily ~/DropboxBackup/db_backup.sh -l loginPathName -d dbName > /dev/null
30 0 * * 1 ~/DropboxBackup/log_backup.sh > /dev/null
```

This cron schedule will execute the `db_backup.sh` script every day at midnight and the `log_backup.sh` script every Monday at 00:30hrs.

> Note: To limit the space consumed on Dropbox, backups are automatically deleted after 28 days.

To list the cron history use:
```bash
$ grep CRON /var/log/syslog
```

### Crontab Syntax

The syntax for adding a new cron job is:
```
1 2 3 4 5 /path/to/command arg1 arg2
```

Where:

1. = Minute (0-59)
2. = Hours (0-23)
3. = Day (0-31)
4. = Month (0-12 [12 == December])
5. = Day of the week (0-7 [7 or 0 == sunday])

The syntax can also be expressed as this:

```
* * * * * command to be executed
- - - - -
| | | | |
| | | | ----- Day of week (0 - 7) (Sunday == 0 or 7)
| | | ------- Month (1 - 12)
| | --------- Day of month (1 - 31)
| ----------- Hour (0 - 23)
------------- Minute (0 - 59)
```

Alternatively, a special string can be used to specify the first five fields:
```
Special string    Meaning
@reboot           Run once, at start-up
@yearly           Run once a year, "0 0 1 1 *"
@annually         (same as @yearly)
@monthly          Run once a month, "0 0 1 * *"
@weekly           Run once a week, "0 0 * * 0"
@daily            Run once a day, "0 0 * * *"
@midnight         (same as @daily)
@hourly           Run once an hour, "0 * * * *"
```

## Sending Email using Mandrill

Linux can be configured to send email through [Mandrill](https://mandrill.com/) using your SMTP & API credentials. The information in this section is based on [this article from Mandrill.com](http://help.mandrill.com/entries/23060367-Can-I-configure-Postfix-to-send-through-Mandrill-).

1. Install a SASL authentication package, [Postfix](http://www.postfix.org/) and mail utilities:
```bash
$ apt-get install -y libsasl2-modules postfix mailutils
```

	During the Postfix setup select `Internet with smarthost`. Enter your fully qualified domain name (`FQDN`) e.g., `example.com` and the SMTP relay host `[smtp.mandrillapp.com]`.

2. Then open the Postfix configuration file and change/append any lines that don't exist:
```bash
$ sudo nano /etc/postfix/main.cf
# Make sure these settings exist (append any that don't):
myhostname = <your-domain-name> e.g., example.com
relayhost = [smtp.mandrillapp.com]
# enable SASL authentication
smtp_sasl_auth_enable = yes
# tell Postfix where the credentials are stored
smtp_sasl_password_maps = hash:/etc/postfix/sasl_passwd 
smtp_sasl_security_options = noanonymous
# use STARTTLS for encryption
smtp_use_tls = yes 
```

3. Create an SMTP username and password file:
```bash
$ sudo nano /etc/postfix/sasl_passwd
# Add your Mandrill username and API key
[smtp.mandrillapp.com] USERNAME:API_KEY
```

4. Create the hash db file for Postfix and lock down access to it:
```bash
$ postmap /etc/postfix/sasl_passwd
$ chmod 0600 /etc/postfix/sasl_passwd /etc/postfix/sasl_passwd.db
```

5. Restart the Postfix service:
```bash
$ service postfix restart
```

6. Finally, test the sending of email through Mandrill:
```bash
$ echo "Test Email." | mail -s "Hello" -a "FROM:YOUR_USERNAME@YOUR_DOMAIN.com" TO_USERNAME@TO_DOMAIN.com
```

7. Check the mail log to make sure that the email has been sent properly:
```bash
$ cat /var/log/mail.log
# Try and find this at the end of the file:
status=sent (250 2.0.0 Ok)
```

## Web Cron (When Your Host Does Not Support Cron Tasks)

It is not always possible to run `cron` or `scheduled tasks`, such as on shared hosting or when using particular server configurations. In these situations, web applications often provide a `web cron` URL, which a host can automatically visit at a scheduled time to perform the same functions as a local cron job. The `web_cron.sh` script has been created to allow a `web cron` URL to be visited and the output downloaded to a daily log file, which is automatically backed up to Dropbox.

### Web Cron Usage

The syntax for running a web cron task is:

```bash
$ ~/DropboxBackup/web_cron.sh -u URL -f FILENAME
```

> Note: `FILENAME` will automatically have the `.log` extension appended to it i.e., `archive` will become `archive.log`.

This script will visit the `URL` and download the output to a single daily log file `FILENAME-YEAR-MONTH-DAY.log`. Any previous days log file is compressed into a `.tar.gz` file and uploaded to Dropbox.

The web cron script can be executed regularly on a machine that does have cron capabilities. For example, to execute the script every hour at five minutes past the hour, add the following to your `crontab` file:
```
5 * * * * ~/DropboxBackup/web_cron.sh -u URL -f FILENAME > /dev/null
```

> Note: To limit the space consumed on Dropbox, backups are automatically deleted after 28 days.

# Dropbox Uploader

Dropbox Uploader is a **BASH** script which can be used to upload, download, delete, list files (and more!) from **Dropbox**, an online file sharing, synchronization and backup service. 

It's written in BASH scripting language and only needs **cURL**.

**Why use this script?**

* **Portable:** It's written in BASH scripting and only needs *cURL* (curl is a tool to transfer data from or to a server, available for all operating systems and installed by default in many linux distributions).
* **Secure:** It's not required to provide your username/password to this script, because it uses the official Dropbox API for the authentication process. 

Please refer to the &lt;Wiki&gt;(https://github.com/andreafabrizi/Dropbox-Uploader/wiki) for tips and additional information about this project. The Wiki is also the place where you can share your scripts and examples related to Dropbox Uploader.

## Features

* Cross platform
* Support for the official Dropbox API
* No password required or stored
* Simple step-by-step configuration wizard
* Simple and chunked file upload
* File and recursive directory download
* File and recursive directory upload
* Shell wildcard expansion (only for upload)
* Delete/Move/Rename/Copy/List files
* Create share link

## Getting started

First, clone the repository using git (recommended):

```bash
git clone https://github.com/andreafabrizi/Dropbox-Uploader/
```

or download the script manually using this command:

```bash
curl "https://raw.githubusercontent.com/andreafabrizi/Dropbox-Uploader/master/dropbox_uploader.sh" -o dropbox_uploader.sh
```

Then give the execution permission to the script and run it:

```bash
 $chmod +x dropbox_uploader.sh
 $./dropbox_uploader.sh
```

The first time you run `dropbox_uploader`, you'll be guided through a wizard in order to configure access to your Dropbox. This configuration will be stored in `~/.dropbox_uploader`.

### Configuration wizard

The configuration wizard is pretty self-explanatory. One thing to notice is that if you choose "App permission", your uploads will end up on Dropbox under an `App/<your_app_name>` folder. To have them stored in another folder, such as in `/dir/`, you'll need to give Dropbox-Uploader permission to all Dropbox files.

## Usage

The syntax is quite simple:

```
./dropbox_uploader.sh COMMAND [PARAMETERS]...

[%%]: Optional param
<%%>: Required param
```

**Available commands:**

* **upload** &lt;LOCAL_FILE/DIR ...&gt; &lt;REMOTE_FILE/DIR&gt;  
Upload a local file or directory to a remote Dropbox folder.  
If the file is bigger than 150Mb the file is uploaded using small chunks (default 4Mb); 
in this case a . (dot) is printed for every chunk successfully uploaded and a * (star) if an error 
occurs (the upload is retried for a maximum of three times).
Only if the file is smaller than 150Mb, the standard upload API is used, and if the -p option is used
the default curl progress bar is displayed during the upload process.  
The local file/dir parameter supports wildcards expansion.

* **download** &lt;REMOTE_FILE/DIR&gt; [LOCAL_FILE/DIR]  
Download file or directory from Dropbox to a local folder

* **delete** &lt;REMOTE_FILE/DIR&gt;  
Remove a remote file or directory from Dropbox

* **move** &lt;REMOTE_FILE/DIR&gt; &lt;REMOTE_FILE/DIR&gt;  
Move or rename a remote file or directory

* **copy** &lt;REMOTE_FILE/DIR&gt; &lt;REMOTE_FILE/DIR&gt;  
Copy a remote file or directory

* **mkdir** &lt;REMOTE_DIR&gt;  
Create a remote directory on DropBox

* **list** [REMOTE_DIR]  
List the contents of the remote Dropbox folder

* **share** &lt;REMOTE_FILE&gt;  
Get a public share link for the specified file or directory
 
* **info**  
Print some info about your Dropbox account

* **unlink**  
Unlink the script from your Dropbox account


**Optional arguments passed before the command:**  
* **-f &lt;FILENAME&gt;**  
Load the configuration file from a specific file

* **-s**  
Skip already existing files when download/upload. Default: Overwrite

* **-d**  
Enable DEBUG mode

* **-q**  
Quiet mode. Don't show progress meter or messages

* **-p**  
Show cURL progress meter

* **-k**  
Doesn't check for SSL certificates (insecure)


**Examples:**
```bash
    ./dropbox_uploader.sh upload /etc/passwd /myfiles/passwd.old
    ./dropbox_uploader.sh upload *.zip /
    ./dropbox_uploader.sh -p download /backup.zip
    ./dropbox_uploader.sh delete /backup.zip
    ./dropbox_uploader.sh mkdir /myDir/
    ./dropbox_uploader.sh upload "My File.txt" "My File 2.txt"
    ./dropbox_uploader.sh share "My File.txt"
    ./dropbox_uploader.sh list
```

## Tested Environments

* GNU Linux
* FreeBSD 8.3/10.0
* MacOSX
* Windows/Cygwin
* Raspberry Pi
* QNAP
* iOS
* OpenWRT
* Chrome OS
* OpenBSD

If you have successfully tested this script on others systems or platforms please let me know!

## Running as cron job
Dropbox Uploader relies on a different configuration file for each system user. The default configuration file location is HOME_DIRECTORY/.dropbox_uploader. This means that if you do the setup with your user and then you try to run a cron job as root, it won't works.  
So, when running this script using cron, please keep in mind the following:
* Remember to setup the script with the user used to run the cron job
* Use always the -f option to specify the full configuration file path, because sometimes in the cron environment the home folder path is not detected correctly
* My advice is, for security reasons, to not share the same configuration file with different users

## How to setup a proxy

To use a proxy server, just set the **https_proxy** environment variable:

**Linux:**
```bash
    export HTTP_PROXY_USER=XXXX
    export HTTP_PROXY_PASSWORD=YYYY
    export https_proxy=http://192.168.0.1:8080
```

**BSD:**
```bash
    setenv HTTP_PROXY_USER XXXX
    setenv HTTP_PROXY_PASSWORD YYYY
    setenv https_proxy http://192.168.0.1:8080
```
   
## BASH and Curl installation

**Debian & Ubuntu Linux:**
```bash
    sudo apt-get install bash (Probably BASH is already installed on your system)
    sudo apt-get install curl
```

**BSD:**
```bash
    cd /usr/ports/shells/bash && make install clean
    cd /usr/ports/ftp/curl && make install clean
```

**Cygwin:**  
You need to install these packages:  
* curl
* ca-certificates
* dos2unix

Before running the script, you need to convert it using the dos2unix command.


**Build cURL from source:**
* Download the source tarball from http://curl.haxx.se/download.html
* Follow the INSTALL instructions

## DropShell

DropShell is an interactive DropBox shell, based on DropBox Uploader:

```bash
DropShell v0.2
The Intractive Dropbox SHELL
Andrea Fabrizi - andrea.fabrizi@gmail.com

Type help for the list of the available commands.

andrea@Dropbox:/$ ls
 [D] 0       Apps
 [D] 0       Camera Uploads
 [D] 0       Public
 [D] 0       scripts
 [D] 0       Security
 [F] 105843  notes.txt
andrea@DropBox:/ServerBackup$ get notes.txt
```

## Donations

 If you want to support this project, please consider donating:
 * PayPal: andrea.fabrizi@gmail.com
 * BTC: 1JHCGAMpKqUwBjcT3Kno9Wd5z16K6WKPqG


CHANGELOG.md
# CHANGELOG Dropbox Backup

## v1.2 [25-Jan-2015]
* [CHANGE] Use `mysql_config_editor` to store MySQL database login credentials in an encrypted format.

## v1.1.3 [29-Dec-2014]
* [CHANGE] Merge upstream changes to `dropbox_uploader.sh`.

## v1.1.2 [26-Oct-2014]
* [CHANGE] Modify the cron schedule for the log backup script to every Monday @ 00:30hrs.
* [CHANGE] Delete the backup log file after being successfully stored in Dropbox.

## v1.1.1 [19-Oct-2014]
* [CHANGE] Use a variable to store the location of the dropbox uploader configuration file.

## v1.1 [16-Oct-2014]
* [FIX] Hard coded temp directory in `log_backup.sh`.
* [NEW] Added a `web_cron.sh` script to allow a web cron URL to be accessed and the output downloaded to a daily log file, which is automatically backed up to Dropbox.

## v1.0 [14-Oct-2014]
* [NEW] Added a MySQL backup script `db_backup.sh`
* [NEW] Added a log file backup script `log_backup.sh`
* [NEW] Automatically delete old backups from Dropbox


# CHANGELOG Dropbox-Uploader

## Version 0.14 - 15 Jul 2014
* Fix issue #112, replaced bash condition -f with -e
* Fix issue #117, sed: 1: invalid command code z
* Fix issue #122, Directory upload doesn't work
* Fix issue #132, enhanced cross compatibility for bash conditions

## Version 0.13 - 17 Jan 2014
* Minor languages fixes, based on issue #88
* Fix issue #91, sed take long time to process regexp on big data
* Fix issue #87, second chunked upload fails after successful first
* Fix issue #97, greadlink not selected on OSX Mavericks
* Fix issues #99 and #100: The stat command in OSX 10.9 has no parameter named --format
* Fix issue #99, clarified how to install greadlink (via homebrew) for Mac OS X. Thanks to paultreny
* Improved list command and minor changes
* Fix issue #103, transform Unicode code-points to characters in list command
* Add VERSION output in DEBUG mode
* DropShell: Fixed issue with quoted command parameters
* Fix issue with absolute paths in CD command
* Enhancement request #108: add initial wildcard support in upload command
* Now the destination file/dir for the upload command is mandatory
* Fixed issue listing files/folders names containing quotes

## Version 0.12.1 - 24 Oct 2013
* Fixed unicode problem with special chars in file names
* Fixed problem removing temp files
* Improved detection of configuration error and better handling of http errors
* Updated setup procedure
* Fixed incompatibility with old bash version

## Version 0.12 - 06 Sep 2013
* Fixed problem uploading twice the same directory
* Add checking for not allowed file names
* Handling connection/resolving host errors
* Better error handling
* Converted if condition to BASH native syntax
* Add normalize_path function, to canonicalize paths
* Add -s option to skip existing files when upload/download
* Removed check for free quota on every file upload to improve performance
* Add checks for file permissions before writing/reading files on the local fs
* Add copy function
* Fixed unicode problem in file listing
* A lot of minor changes and improvements

## Version 0.11.9 - 27 Jun 2013
* Add missing -f option in usage
* Add some /dev/null redirects
* Improved error messages
* Improved grep syntax
* Add function to check the cURL exit code
* Improved messages output
* Add new command line options: -q -d -k

## Version 0.11.8 - 05 Jun 2013
* Add move/rename function
* Improved the configuration file management (thanks to Robert G.)
* Updated strings to reflect the new Dropbox "Create App" page
* Add support for download directories
* Add support for upload directories

## Version 0.11.7 - 23 Apr 2013
* Fixed issue with special chars
* Fix for iOS

## Version 0.11.6 - 15 Mar 2013
* Add optional command-line parameter ('-f') to read dropbox configuration from a specific file (thanks to pjv)

## Version 0.11.5 - 22 Gen 2013
* Added the ability to get a share link for a specified file (thanks to camspiers)

## Version 0.11.4 - 17 Gen 2013
* Fix for QNAP compatibility (thanks to Fritz Ferstl)
* Implemented mkdir command (thanks to Joel Maslak)
* Fix for Solaris compatibility

## Version 0.11.3 - 22 Dec 2012:
* Improved list command (thanks to Robert González)
* Fixed problem with unicode characters

## Version 0.11.2 - 14 Nov 2012:
* Added a check for the free quota before uploading a file
* Now the quota informations are displayed in Mb
* Removed urlencode function for incompatibility with older curl versions
* Fixed problem uploading files that contains @ character
* Minor changes

## Version 0.11.1 - 12 Nov 2012:
* As suggested by the DropBox API documentation, the default chunk for chunked uploads is now 4Mb
* Minor changes

## Version 0.11 - 11 Nov 2012:
* Parameterized the curl binary location
* Fix for MacOSX 10.8 (thanks to Ben - www.aquiltforever.com)

## Version 0.10 - 03 Nov 2012:
* Code clean
* Improved urlencode function (thanks to Stefan Trauth * www.stefantrauth.de)
* Added command remove as alias of delete
* Fix for Raspberry PI
* Now if an error occurs during a chunk uploading, the upload is retried for a maximum of three times
* Minor changes
* Tested on Cygwin and MacOSX

## Version 0.9.9 - 24 Oct 2012:
* Added the possibility to choose the access level (App folder o Full Dropbox) during the setup procedure
* Added a check for the BASH shell version
* Fixed problems in listing files/directories with special characters
* Added the option CURL_ACCEPT_CERTIFICATES (see the script source)
* Added back the standard upload function. Now only if the file is greater than 150Mb, the chunked_upload API will be used.
* Fixed compatibility with bsd sed. Tested on FreeBSD, but probably it works on others bsd versions and osx. Let me know!
* Minor changes

## Version 0.9.8 - 03 Oct 2012:
* Implemented chunked upload. Now there is no limit to file size!

## Version 0.9.7 - 14 Sep 2012:
* Fixed bug in listing empty directories

## Version 0.9.6 - 12 Sep 2012:
* Implemented list command
* Minor changes

## Version 0.9.5 - 18 Jul 2012:

web_cron.sh
#!/bin/bash

########################################################
# A script to download and backup the output from a web 
# cron URL.
#
# Author: Dan @ ClickBeyond
# Website: http://clickbeyond.github.io/DropboxBackup/
# Copyright 2015
#
# Usage: $ ./web_cron.sh -u URL -f FILENAME
#
# Before running this script for the first time make 
# sure you have already run the `./dropbox_uploader.sh` 
# script. Also make sure you give this script execution 
# permissions:
# $ chmod +x web_cron.sh
#
########################################################

# Exit immediately if an error occurs (i.e., a command exits with a non-zero status)
set -e

usage() { echo "Usage: $0 [-u <url>] [-f <filename>]" 1>&2; exit 1; }

while getopts ":u:f:" o; do
    case "${o}" in
        u)
            url=${OPTARG}
            ;;
        f)
            filename=${OPTARG}
            ;;
        *)
            usage
            ;;
    esac
done
shift $((OPTIND-1))

if [ -z "${url}" ] || [ -z "${filename}" ]; then
    usage
fi

# Save the web cron URL output to a daily log file
NOW=$(date +"%Y-%m-%d")
LOG_FILE="$filename-$NOW.log"
TMP_DIR="tmp"

cd ~
# Create a temp directory if it doesnt exist
mkdir -p $TMP_DIR
echo "Starting URL backup for '$url' to $LOG_FILE"
start=$SECONDS
wget --quiet $url -O - >> ~/$TMP_DIR/$LOG_FILE
duration=$(( SECONDS - start ))
echo "Archiving complete! Finished in $duration seconds!"

# Backup the previous days archive to Dropbox if it exists
DROP_DIR="Archive_Backups"
DROPBOX_UPLOADER=~/DropboxBackup/dropbox_uploader.sh
DROP_CONFIG=~/.dropbox_uploader
BKP_DATE=$(date --date="-1 day" +%Y-%m-%d)
PREV_DAY_LOG_FILE="$filename-$BKP_DATE.log"
BKP_LOG_FILE="$filename-backup-$BKP_DATE.tar.gz"

# Set the delete date to a multiple of the cron schedule e.g., a weekly cron schedule means data can only be deleted every 7, 14, 21, 28, etc, days
DEL_DATE=$(date --date="-28 day" +%Y-%m-%d)

if [ -f ~/$TMP_DIR/$PREV_DAY_LOG_FILE ];
then
	echo "Starting archive backup of '$PREV_DAY_LOG_FILE' to Dropbox..."
	start=$SECONDS
	cd ~/$TMP_DIR
	tar -zcf "$BKP_LOG_FILE" "$PREV_DAY_LOG_FILE"
	$DROPBOX_UPLOADER -f $DROP_CONFIG upload $BKP_LOG_FILE "/$DROP_DIR/$BKP_LOG_FILE"
	echo "Deleting '$PREV_DAY_LOG_FILE' and '$BKP_LOG_FILE'"
	rm -f $PREV_DAY_LOG_FILE $BKP_LOG_FILE
	
	# Delete old backup from Dropbox
	echo "Finding any existing archive backup that was made on '$DEL_DATE'..."
	while read -r state size file rest
	do
		if [[ $state = "[F]" && $file = "$filename-backup-"* ]]
		then
			if [[ $file = "$filename-backup-$DEL_DATE"* ]]
			then
				echo "Old backup found! Deleting '$file'"
				$DROPBOX_UPLOADER -f $DROP_CONFIG delete "/$DROP_DIR/$file"
			fi
		fi
	done < <($DROPBOX_UPLOADER -f $DROP_CONFIG list $DROP_DIR/)
	
	duration=$(( SECONDS - start ))
	echo "Archive backup complete! Finished in $duration seconds!"
else
	echo "File '$PREV_DAY_LOG_FILE' does not exist! Skipping backup to Dropbox..."
fi


db_backup.sh
#!/bin/bash

#############################################################################################
# A script to backup your MySQL database to dropbox.
#
# Author: Dan @ ClickBeyond
# Website: http://clickbeyond.github.io/DropboxBackup/
# Copyright 2015
#
# Usage: $ ./db_backup.sh -l loginPathName -d dbName
#
# Before running this script for the first time make sure you execute the following commands:
# 1. $ chmod +x dropbox_uploader.sh
# 2. $ chmod +x db_backup.sh
# 3. $ ./dropbox_uploader.sh
# The last step will configure access to your dropbox account if this is the first time you
# have run dropbox_uploader.sh.
#
# Based on: http://robinadr.com/2013/10/automated-database-backups-to-dropbox
#############################################################################################

# Exit immediately if an error occurs (i.e., a command exits with a non-zero status)
set -e

usage() { echo "Usage: $0 [-l <loginPathName>] [-d <dbName>]" 1>&2; exit 1; }

while getopts ":l:d:" o; do
    case "${o}" in
        l)
            dbLogin=${OPTARG}
            ;;
		d)
            dbName=${OPTARG}
            ;;
        *)
            usage
            ;;
    esac
done
shift $((OPTIND-1))

if [ -z "${dbLogin}" ] || [ -z "${dbName}" ]; then
    usage
fi

TMP_DIR="tmp"
DROP_DIR="MySQL_Backups"
NOW=$(date +"%Y-%m-%d_%H-%M-%S")
SQL_FILE="backup-$dbName-$NOW.sql"
BKP_FILE="backup-$dbName-$NOW.tar.gz"
#BKP_DIRS="./dir/to/backup"
DROPBOX_UPLOADER=~/DropboxBackup/dropbox_uploader.sh
DROP_CONFIG=~/.dropbox_uploader
LOG_FILE=~/$TMP_DIR/backup.log
# Set the delete date to a multiple of the cron schedule e.g., a weekly cron schedule means data can only be deleted every 7, 14, 21, 28, etc, days
DEL_DATE=$(date --date="-28 day" +%Y-%m-%d)

# Define a date function for adding a timestamp to the log file
adddate() {
    while IFS= read -r line; do
        echo "$(date) $line"
    done
}

cd ~
# Create temp directory if it doesnt exist
mkdir -p $TMP_DIR
echo "Starting backup of the '$dbName' MySQL database to '$SQL_FILE'" | adddate >> $LOG_FILE
start=$SECONDS
cd $TMP_DIR
mysqldump --login-path=$dbLogin -B $dbName > $SQL_FILE
tar -zcf "$BKP_FILE" $SQL_FILE

#echo "Backing up the directories $BKP_DIRS"
#tar -zcf "$BKP_FILE" $BKP_DIRS $SQL_FILE

echo "Uploading to Dropbox..." | adddate >> $LOG_FILE
$DROPBOX_UPLOADER -f $DROP_CONFIG upload $BKP_FILE "/$DROP_DIR/$BKP_FILE" | adddate >> $LOG_FILE

rm -f $BKP_FILE $SQL_FILE

# Delete old backup from Dropbox
echo "Finding any existing backup that was made on '$DEL_DATE'..." | adddate >> $LOG_FILE
while read -r state size file rest
do
    if [[ $state = "[F]" && $file = "backup-$dbName-"* ]]
    then
        if [[ $file = "backup-$dbName-$DEL_DATE"* ]]
        then
			echo "Old backup found! Deleting '$file'" | adddate >> $LOG_FILE
			$DROPBOX_UPLOADER -f $DROP_CONFIG delete "/$DROP_DIR/$file" | adddate >> $LOG_FILE
        fi
    fi
done < <($DROPBOX_UPLOADER -f $DROP_CONFIG list $DROP_DIR/)

duration=$(( SECONDS - start ))
echo "Backup complete! Finished in $duration seconds!" | adddate >> $LOG_FILE

log_backup.sh
#!/bin/bash

###################################################
# A script to backup your log file to dropbox.
#
# Author: Dan @ ClickBeyond
# Website: http://clickbeyond.github.io/DropboxBackup/
# Copyright 2015
#
# Usage: $ ./log_backup.sh
#
# Before running this script for the first time make 
# sure you have already run the `./db_backup.sh` script. 
# Also make sure you give this script execution 
# permissions:
# $ chmod +x log_backup.sh
#
###################################################

# Exit immediately if an error occurs (i.e., a command exits with a non-zero status)
set -e

TMP_DIR="tmp"
DROP_DIR="Log_Backups"
LOG_FILE="backup.log"
DROPBOX_UPLOADER=~/DropboxBackup/dropbox_uploader.sh
DROP_CONFIG=~/.dropbox_uploader
# Set the delete date to a multiple of the cron schedule e.g., a weekly cron schedule means data can only be deleted every 7, 14, 21, 28, etc, days
DEL_DATE=$(date --date="-28 day" +%Y-%m-%d)

if [ -f ~/$TMP_DIR/$LOG_FILE ];
then
	echo "Starting backup of 'backup.log' to Dropbox..."
	start=$SECONDS
	cd ~/$TMP_DIR
	BKP_LOG_FILE="log-backup-$(date +"%Y-%m-%d_%H-%M-%S").tar.gz"
	tar -zcf "$BKP_LOG_FILE" "backup.log"
	$DROPBOX_UPLOADER -f $DROP_CONFIG upload $BKP_LOG_FILE "/$DROP_DIR/$BKP_LOG_FILE"
	rm -f $BKP_LOG_FILE $LOG_FILE
	
	# Delete old backup from Dropbox
	echo "Finding any existing backup that was made on '$DEL_DATE'..."
	while read -r state size file rest
	do
		if [[ $state = "[F]" && $file = "log-backup-"* ]]
		then
			if [[ $file = "log-backup-$DEL_DATE"* ]]
			then
				echo "Old backup found! Deleting '$file'"
				$DROPBOX_UPLOADER -f $DROP_CONFIG delete "/$DROP_DIR/$file"
			fi
		fi
	done < <($DROPBOX_UPLOADER -f $DROP_CONFIG list $DROP_DIR/)
	
	duration=$(( SECONDS - start ))
	echo "Log backup complete! Finished in $duration seconds!"
else
	echo "File '~/$TMP_DIR/$LOG_FILE' does not exist! Please run './db_backup.sh' before running this script again."
fi
